torch/csrc/distributed/c10d/ProcessGroupCudaP2P.cpp

```diff
@@ -0,0 +1,206 @@
+#ifdef USE_C10D_NCCL
+#include <torch/csrc/distributed/c10d/ProcessGroupCudaP2P.hpp>
+
+#include <torch/csrc/distributed/c10d/PrefixStore.hpp>
+
+namespace c10d {
+
+using namespace c10d::intra_node_comm;
+
+ProcessGroupCudaP2P::ProcessGroupCudaP2P(
+    const c10::intrusive_ptr<Store>& store,
+    int rank,
+    int size,
+    c10::intrusive_ptr<Options> options)
+    : Backend(rank, size), stream_(c10::cuda::getStreamFromPool()) {
+  nccl_backend_ = c10::make_intrusive<ProcessGroupNCCL>(
+      c10::make_intrusive<PrefixStore>("nccl", store),
+      rank,
+      size,
+      options->nccl_options);
+  nccl_backend_->setSequenceNumberForGroup();
+
+  p2p_backend_ = c10::make_intrusive<IntraNodeComm>(
+      c10::make_intrusive<PrefixStore>("p2p", store),
+      rank,
+      size,
+      options->buffer_size);
+  if (!p2p_backend_->rendezvous()) {
+    p2p_backend_ = nullptr;
```

Comment:
shall we throw an error here or you want to preserve the grace fallback to NCCL?

Reply:
Yes the intention is to gracefully fallback. `p2p_backend_` being `nullptr` will make `is_p2p_available` return false, which instructs the application layer to fallback. But I also see benefits of just erroring out here. Let me think more.